### ISSUE TITLE
better diagnostic for implicit Box<Trait + 'static>

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#31812]
[This](http://is.gd/A3Uyx7) doesn't compile because `Box<Trait>` means `Box<Trait + 'static>`. It would be nice if there was a note to that effect, because otherwise it's not obvious at all where the lifetime bound comes from.

**Comment by GuillaumeGomez** [comment]
What's even more strange is that it doesn't span an error code as well. As far as I know, libborrowck is built after libsyntax so it should have error codes. The code which spans these messages is here `https://github.com/rust-lang/rust/blob/master/src/librustc_borrowck/borrowck/mod.rs#L1035-L1039`. I'm going to take to see why there is no error code in here.

**Comment by arielb1** [comment]
This is not specific to trait objects: borrowck errors don't report a lifetime origin - the lifetime origin is not available at the stage borrowck runs.

For example:

```Rust
struct Foo(&'static u32);

fn main() {
    let x = 42;
    Foo(&x);
}
```

**Comment by arielb1** [comment]
@GuillaumeGomez 

borrowck is going to be rewritten pretty soon.

**Comment by GuillaumeGomez** [comment]
I'll check once it'll be done then.

**Comment by durka** [comment]
I'll close this for now then as it seems impractical to fix at the moment, until the borrowck architecture changes.
